From cd10091f03f6255a47d7146eea5738f1f4ceea35 Mon Sep 17 00:00:00 2001
From: Vignesh Viswanathan <viswanat@codeaurora.org>
Date: Tue, 12 Sep 2017 15:21:16 +0530
Subject: qcacld-3.0: Fix out-of-bounds access in
 lim_process_action_vendor_specific

Currently in the function lim_process_action_vendor_specific, mem_cmp
is done on the received frame pointer without validating the frame_len
which could lead to out-of-bounds memory access if the frame_len is
not matching the size of action_hdr.

Add check to validate the frame_len with action_hdr size before doing
mem_cmp for the p2p oui.

Change-Id: I39329d1a9ef45614d3c617db11a7a7f5ec2aaaec
CRs-Fixed: 2101439
---
 core/mac/src/pe/lim/lim_process_action_frame.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/core/mac/src/pe/lim/lim_process_action_frame.c b/core/mac/src/pe/lim/lim_process_action_frame.c
index fa7ea90..7b94dc8 100644
--- a/core/mac/src/pe/lim/lim_process_action_frame.c
+++ b/core/mac/src/pe/lim/lim_process_action_frame.c
@@ -1639,6 +1639,11 @@ static void lim_process_action_vendor_specific(tpAniSirGlobal mac_ctx,
 
 	mac_hdr = WMA_GET_RX_MAC_HEADER(pkt_info);
 	frame_len = WMA_GET_RX_PAYLOAD_LEN(pkt_info);
+	if (frame_len < sizeof(action_hdr)) {
+		pe_debug("Received action frame of invalid len %d", frame_len);
+		return;
+	}
+
 	if (session)
 		session_id = session->smeSessionId;
 	/* Check if it is a P2P or DPP public action frame. */
-- 
cgit v1.1

